### PR TITLE
On 752 get learner details

### PR DESCRIPTION
**Pull request by martin-hornsby**

### CODE REVIEW

File: src/SFA.DAS.AssessorService.Application.Api.External.UnitTests/Controllers/Learner/LearnerTestBase.cs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        }
+
+        protected void SetFakeHttpMessageHandlerResponse(HttpStatusCode status, object content)
+        {
+            FakeHttpMessageHandlerMock.Setup(f => f.Send(It.IsAny<HttpRequestMessage>())).Returns(
+                new HttpResponseMessage
+                {
+                    StatusCode = status,
+                    Content = new StringContent(JsonConvert.SerializeObject(content))
+                });
```
Comment (davidgouge) [inline src/SFA.DAS.AssessorService.Application.Api.External.UnitTests/Controllers/Learner/LearnerTestBase.cs]: Have a look at using https://github.com/richardszalay/mockhttp to easily mock HttpClient. We use it elsewhere in the solution. 👍

File: src/SFA.DAS.AssessorService.Application.Api.External/Controllers/CertificateController.cs [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+                    certificateResponse.Certificate = updateCertificateResponse.Certificate;
+                    certificateResponse.Status = updateCertificateResponse.Status;
+                    certificateResponse.ValidationErrors.AddRange(updateCertificateResponse.ValidationErrors);
+                }
+
+                response.Add(certificateResponse);
+            }
+
+            return Ok(response);
+        }
```
Comment (davidgouge) [inline src/SFA.DAS.AssessorService.Application.Api.External/Controllers/CertificateController.cs]: These controllers are doing too much really.  Push as much as you can back into the handler in the internal API.  Both the external and internal API layers should be as thin and dumb as possible. 😁

File: src/SFA.DAS.AssessorService.Application.Api.External/Controllers/LearnerController.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+                Uln = uln,
+                Surname = lastname,
+                UkPrn = _headerInfo.Ukprn,
+                Username = _headerInfo.Username
+            };
+
+            List<SearchResult> results = await _apiClient.Search(searchQuery, stdCode);
+            
+            return Ok(results);
+        }
```
Comment (davidgouge) [inline src/SFA.DAS.AssessorService.Application.Api.External/Controllers/LearnerController.cs]: This one's perfect as it just gathers the info it needs and shoves it off the to the Internal API. 👍

File: src/SFA.DAS.AssessorService.Application.Api.External/Infrastructure/ApiClient.cs [hunk 4]
```diff
@@ -15,23 +22,81 @@ public class ApiClient
...
+                {
+                    return await Task.Factory.StartNew<U>(() => JsonConvert.DeserializeObject<U>(json, _jsonSettings));
+                }
+                else
+                {
+                    _logger.LogInformation($"HttpRequestException: Status Code: {response.StatusCode} Body: {json}");
+                    throw new HttpRequestException(json);
+                }
+            }
+        }
```
Comment (davidgouge) [inline src/SFA.DAS.AssessorService.Application.Api.External/Infrastructure/ApiClient.cs]: This looks like it's copied from the old ApiClient project.  Have a look at how I've replaces the Put / Post in the ON-674... branch.

File: src/SFA.DAS.AssessorService.Application.Api.External/Middleware/ErrorHandlingMiddleware.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+                var response = new ApiResponse(context.Response.StatusCode, ex.Message);
+                var json = JsonConvert.SerializeObject(response,
+                    new JsonSerializerSettings
+                    {
+                        ContractResolver = new CamelCasePropertyNamesContractResolver()
+                    });
+
+                await context.Response.WriteAsync(json);
+            }
+        }
```
Comment (davidgouge) [inline src/SFA.DAS.AssessorService.Application.Api.External/Middleware/ErrorHandlingMiddleware.cs]: Get rid of this error handling middleware for the time being. It's a pain for dev and we prob won't need it anyway.

File: src/SFA.DAS.AssessorService.Application.Api.External/Middleware/GetHeadersMiddleware.cs [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+            context.Request.Headers.TryGetValue("x-username", out var usernameHeaderValue);
+            context.Request.Headers.TryGetValue("x-ukprn", out var ukprnHeaderValue);
+
+            int.TryParse(ukprnHeaderValue.FirstOrDefault(), out int ukprn);
+            string username = usernameHeaderValue.FirstOrDefault() ?? string.Empty;
+
+            headerInfo.Ukprn = ukprn;
+            headerInfo.Username = username;
+
+            await _next(context);
```
Comment (davidgouge) [inline src/SFA.DAS.AssessorService.Application.Api.External/Middleware/GetHeadersMiddleware.cs]: Can you get this to return a BadRequest here if those headers aren't present.

File: src/SFA.DAS.AssessorService.Application.Api.External/Models/Search/SearchQuery.cs [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+                const int multiplier = 16777619;
+
+                int hash = hashBase;
+                hash = (hash * multiplier) ^ Uln.GetHashCode();
+                hash = (hash * multiplier) ^ (Surname is null ? 0 : Surname.GetHashCode());
+                hash = (hash * multiplier) ^ UkPrn.GetHashCode();
+                hash = (hash * multiplier) ^ (Username is null ? 0 : Username.GetHashCode());
+                return hash;
+            }
+        }
```
Comment (davidgouge) [inline src/SFA.DAS.AssessorService.Application.Api.External/Models/Search/SearchQuery.cs]: Are you using any of this equality stuff? If not, get rid for now until needed. 👍

File: src/SFA.DAS.AssessorService.Application.Api.External/Infrastructure/ApiClient.cs (hunk 4 above)
Comment (martin-hornsby) [reply src/SFA.DAS.AssessorService.Application.Api.External/Infrastructure/ApiClient.cs]: I checked that branch and it looks the same and same as Web.Staff . Are you expecting me to use  SFA.DAS.AssessorService.Application.Api.Client.Clients.ApiClientBase ?

File: src/SFA.DAS.AssessorService.Application.Api.External/Models/Search/SearchQuery.cs (hunk 7 above)
Comment (martin-hornsby) [reply src/SFA.DAS.AssessorService.Application.Api.External/Models/Search/SearchQuery.cs]: Yeah I'm using it for my Unit Test Asserts . It's similar to what ReSharper generates.

**Comment by davidgouge** [comment]
Existing EPAO app tested as working as expected on this branch.  Merging to master.

**Pull request merged**